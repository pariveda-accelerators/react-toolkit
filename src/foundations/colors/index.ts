export * from './color-theory';
export * from './sass-palette';
export * from './understanding-hex';
export * from './creating-color';
export * from './color-and-brand';
export * from './brand-demo';
export * from './colors';

export const colors = [
  'white',
  'black',
  'red',
  'orange',
  'yellow',
  'lime',
  'green',
  'seafoam',
  'cyan',
  'azure',
  'blue',
  'violet',
  'magenta',
  'rose',
];
export const shades = ['10', '20', '30', '40', '50', '60', '70', '80', '90', '100'];

export type TCcolorName = typeof colors[number];
export type TShade = typeof shades[number];

export interface IColor {
  color: TColor;
  shade: TShade;
}

export type TColor = TCcolorName | IColor;
