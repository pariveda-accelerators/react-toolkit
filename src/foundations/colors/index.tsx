import * as React from 'react';
import styled from 'styled-components';

interface IColorSwatch {
  color: string;
}

const ColorSwatch = styled.div<IColorSwatch>`
  height: 64px;
  width: 64px;
  border-radius: 4px;
  background-color: var(${props => `--${props.color}`});
`;

interface IFlex {
  row?: boolean;
  column?: boolean;
  padding?: number;
}
const Flex = styled.div<IFlex>`
  display: flex;
  flex-direction: ${props => (props.column ? 'column' : 'row')};
  padding: ${props => props.padding || '8px'};
  align-items: center;
  font-family: sans-serif;
`;

const cssNames = ['grey', 'red', 'orange', 'yellow', 'green', 'blue', 'purple'];

export const ColorPalette = () => {
  return (
    <Flex column>
      {cssNames.map((color, colorIdx) => (
        <Flex key={colorIdx}>
          <ColorSwatch color={`${color}-1`} />
          <Flex />
          <ColorSwatch color={`${color}-2`} />
          <Flex />
          <ColorSwatch color={`${color}-3`} />
          <Flex />
          <ColorSwatch color={`${color}-4`} />
          <Flex />
          <ColorSwatch color={`${color}-5`} />
          <Flex />
          <ColorSwatch color={`${color}-6`} />
          <Flex />
          <ColorSwatch color={`${color}-7`} />
        </Flex>
      ))}
    </Flex>
  );
};
