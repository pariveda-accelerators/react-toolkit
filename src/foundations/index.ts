export * from './typography';
export * from './layout';
