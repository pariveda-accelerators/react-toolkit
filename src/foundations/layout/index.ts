export * from './box';
export * from './block';
export * from './flex';
export * from './grid';
export * from './grids';
