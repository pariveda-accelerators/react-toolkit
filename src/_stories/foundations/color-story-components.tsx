import React from 'react';
import { Box, Column, Row } from 'foundations/layout';
import { COLOR } from 'foundations/colors';
import { Story, ColorStack, TRenderColorShade } from '_stories/story-components';

const renderColorShade: TRenderColorShade = ({ colorShade }) => (
  <Box bg={colorShade} />
);

export const Colors = () => {
  return (
    <Story title="Colors">
      <Column>
        {COLOR.slice(3).map(color => (
          <Row key={color} p="s">
            <ColorStack
              color={color}
              propName="bg"
              renderColorShade={renderColorShade}
              separateHeadingFromContent
            />
          </Row>
        ))}
      </Column>
    </Story>
  );
};
