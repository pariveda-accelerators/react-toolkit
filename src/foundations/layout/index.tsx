import React, { FC } from 'react';
import { IColorObject, getPaletteColor } from 'foundations/colors';
import styled from 'styled-components';

interface IBox {
  className?: string;
  column?: boolean;
  bgColor?: IColorObject;
  p?: string;
}

const Box: FC<IBox> = props => {
  return <div className={props.className}>{props.children}</div>;
};

const getSharedProps = (props: IBox) => {
  const bgColor = props.bgColor
    ? `var(--${getPaletteColor(props.bgColor)})`
    : 'transparent';
  return `
    background-color: ${bgColor};
    padding: ${props.p ? props.p : 0};
  `;
};

export const Flex = styled(Box)`
  display: flex;
  flex-direction: ${props => (props.column ? 'column' : 'row')};
  ${props => getSharedProps(props)}
`;

export const Block = styled(Box)`
  display: block;
  ${props => getSharedProps(props)}
`;
