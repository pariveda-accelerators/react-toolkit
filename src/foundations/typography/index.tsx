import React, { FC } from 'react';
import classnames from 'classnames';
import { createClassName, isNullOrUndefined } from '../../utilities';
import { IDefaultProps, getDefaultProps } from '../../types';
import { TColorShade } from 'foundations/colors';
import { TShirtSize } from 'foundations/layout/box';

/**
 * Docs on everything you can do w/ fonts: https://developer.mozilla.org/en-US/docs/Learn/CSS/Styling_text/Fundamentals
 *
 * Future support is needed for:
 *   * text-overflow
 *   * text-indent
 *   * white-space
 *   * word-break
 */

//#region Font Tag
export const FONT_TAG = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p', 'label'] as const;
export type TFontTag = typeof FONT_TAG[number];
//#endregion Font Tag
//#region Font Type
export const FONT_TYPE = [
  'display',
  'title',
  'subtitle',
  'section-title',
  'section-subtitle',
  'subsection-title',
  'body',
  'description',
  'label',
] as const;
export type TFontType = typeof FONT_TYPE[number];
//#endregion Font Type
//#region Font Weight
export const FONT_WEIGHT = ['normal', 'light', 'bold', 'strong'] as const;
export type TFontWeight = typeof FONT_WEIGHT[number];
//#endregion Font Weight
//#region Font Style
export const FONT_STYLE = ['italics', 'emphasis'] as const;
export type TFontStyle = typeof FONT_STYLE[number];
//#endregion Font Style
//#region Color & Accessibility Tags
export interface ChildrenAsReactNode {
  children: React.ReactNode;
}
const Emphasis: FC<ChildrenAsReactNode> = props => <em {...props} />;
const Italics: FC<ChildrenAsReactNode> = props => <i {...props} />;
const Strong: FC<ChildrenAsReactNode> = props => <strong {...props} />;
const Bold: FC<ChildrenAsReactNode> = props => <b {...props} />;
const wrapChildren = (
  Em: FC<ChildrenAsReactNode>,
  B: FC<ChildrenAsReactNode>,
  children: React.ReactNode
) => (
  <Em>
    <B>{children}</B>
  </Em>
);
//#endregion Style & Accessibility Tags
//#region Font Align
export const TEXT_ALIGN = ['left', 'center', 'right'] as const;
export type TTextAlign = typeof TEXT_ALIGN[number];
//#endregion Font Align
//#region Text Transform
export const TEXT_TRANSFORM = [
  'none',
  'uppercase',
  'lowercase',
  'capitalize',
] as const;
export type TTextTransform = typeof TEXT_TRANSFORM[number];
//#endregion Text Transform

interface IBaseFont extends IDefaultProps {
  /** Font Tag */
  tag?: TFontTag;
  /** Font Type */
  type?: TFontType;
  /** Font Weight */
  weight?: TFontWeight;
  /** Font Style */
  fStyle?: TFontStyle;
  /** Text Align */
  align?: TTextAlign;
  /** Text Transform */
  transform?: TTextTransform;
  /** Color */
  color?: TColorShade;
  /** Margin */
  m?: TShirtSize;
}

const BaseFont: FC<IBaseFont> = ({
  tag,
  type,
  weight,
  fStyle,
  align,
  transform,
  color,
  m,
  className,
  children,
  ...props
}) => {
  const defaultProps = getDefaultProps(props);
  const Em =
    fStyle === 'italics'
      ? Italics
      : fStyle === 'emphasis'
      ? Emphasis
      : React.Fragment;
  const B = weight === 'bold' ? Bold : weight === 'strong' ? Strong : React.Fragment;

  const blockStyles = {
    align,
    transform,
    type,
    color,
  };
  const blockClasses = createClassName(blockStyles, { block: 'text' });

  const margin = !isNullOrUndefined(m) ? `m${m}` : undefined;

  const TagElement = React.createElement(
    tag || 'p',
    {
      className: classnames(...blockClasses, margin, className),
      tag,
      ...defaultProps,
    },
    wrapChildren(Em, B, children)
  );
  return TagElement;
};
BaseFont.defaultProps = {
  tag: 'p',
  type: 'body',
  weight: undefined,
  fStyle: undefined,
  align: undefined,
  transform: undefined,
  color: 'black',
};

type TFont = Omit<IBaseFont, 'type'>;
export const Display: FC<TFont> = props => (
  <BaseFont tag="h1" {...props} type="display" />
);
export const Title: FC<TFont> = props => (
  <BaseFont tag="h2" {...props} type="title" />
);
export const Subtitle: FC<TFont> = props => (
  <BaseFont tag="h3" {...props} type="subtitle" />
);
export const SectionTitle: FC<TFont> = props => (
  <BaseFont tag="h4" {...props} type="section-title" />
);
export const SectionSubtitle: FC<TFont> = props => (
  <BaseFont tag="h5" {...props} type="section-subtitle" />
);
export const SubsectionTitle: FC<TFont> = props => (
  <BaseFont tag="h6" {...props} type="subsection-title" />
);
export const Body: FC<TFont> = props => <BaseFont tag="p" {...props} type="body" />;
export const Description: FC<TFont> = props => (
  <BaseFont tag="p" {...props} type="description" />
);
export const Label: FC<TFont> = props => (
  <BaseFont tag="label" {...props} type="body" />
);

export const Fonts: {
  [key: string]: React.ComponentType<TFont>;
} = {
  Display,
  Title,
  Subtitle,
  SectionTitle,
  SectionSubtitle,
  SubsectionTitle,
  Body,
  Description,
  Label,
};
