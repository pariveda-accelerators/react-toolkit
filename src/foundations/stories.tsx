import React from 'react';
import { Colors } from './colors/story-components';
import { Typography } from './typography/story-components';
import { Layout } from './layout/story-components';

export default {
  title: 'Foundations',
};

export const colors = () => <Colors />;
export const typography = () => <Typography />;
export const layout = () => <Layout />;
