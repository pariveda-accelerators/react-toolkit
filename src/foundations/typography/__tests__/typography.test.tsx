import React from 'react';
import { Fonts } from 'foundations';
import { render } from '@testing-library/react';

describe('Typography', () => {
  it('Renders Default', () => {
    const { baseElement, getByText } = render(
      <>
        <Fonts.Display>Display Font</Fonts.Display>
        <Fonts.Title>Title Font</Fonts.Title>
        <Fonts.Subtitle>Subtitle Font</Fonts.Subtitle>
        <Fonts.SectionTitle>SectionTitle Font</Fonts.SectionTitle>
        <Fonts.SectionSubtitle>Section Subtitle Font</Fonts.SectionSubtitle>
        <Fonts.Body>Body Font</Fonts.Body>
        <Fonts.Description>Description Font</Fonts.Description>
        <Fonts.Label>Label Font</Fonts.Label>
      </>
    );

    expect(baseElement).toMatchSnapshot();
    expect(getByText('Display Font').tagName).toBe('H1');
    expect(getByText('Title Font').tagName).toBe('H2');
    expect(getByText('Subtitle Font').tagName).toBe('H3');
    expect(getByText('SectionTitle Font').tagName).toBe('H4');
    expect(getByText('Section Subtitle Font').tagName).toBe('H5');
    expect(getByText('Body Font').tagName).toBe('P');
    expect(getByText('Description Font').tagName).toBe('P');
    expect(getByText('Label Font').tagName).toBe('H6');
  });

  it('Handles Tag', () => {
    const { baseElement, getByText } = render(
      <>
        <Fonts.Display h1>H1 Tag</Fonts.Display>
        <Fonts.Display h2>H2 Tag</Fonts.Display>
        <Fonts.Display h3>H3 Tag</Fonts.Display>
        <Fonts.Display h4>H4 Tag</Fonts.Display>
        <Fonts.Display h5>H5 Tag</Fonts.Display>
        <Fonts.Display h6>H6 Tag</Fonts.Display>
        <Fonts.Display p>P Tag</Fonts.Display>
      </>
    );

    expect(baseElement).toMatchSnapshot();
    expect(getByText('H1 Tag').tagName).toBe('H1');
    expect(getByText('H2 Tag').tagName).toBe('H2');
    expect(getByText('H3 Tag').tagName).toBe('H3');
    expect(getByText('H4 Tag').tagName).toBe('H4');
    expect(getByText('H5 Tag').tagName).toBe('H5');
    expect(getByText('H6 Tag').tagName).toBe('H6');
    expect(getByText('P Tag').tagName).toBe('P');
  });

  it('Handles Emphasis', () => {
    const { baseElement } = render(
      <Fonts.Display emphasis>Emphasis Display Font</Fonts.Display>
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('Handles Italics', () => {
    const { baseElement } = render(
      <Fonts.Display italics>Italics Display Font</Fonts.Display>
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('Handles Strong', () => {
    const { baseElement } = render(
      <Fonts.Display strong>Strong Display Font</Fonts.Display>
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('Handles Bold', () => {
    const { baseElement } = render(
      <Fonts.Display bold>Bold Display Font</Fonts.Display>
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('Handles Alignment', () => {
    const { getByText } = render(
      <>
        <Fonts.Display left>Left Align</Fonts.Display>
        <Fonts.Display center>Center Align</Fonts.Display>
        <Fonts.Display right>Right Align</Fonts.Display>
      </>
    );

    expect(getByText('Left Align')).toHaveClass('text--left');
    expect(getByText('Center Align')).toHaveClass('text--center');
    expect(getByText('Right Align')).toHaveClass('text--right');
  });
  it('Handles Transform', () => {
    const { getByText } = render(
      <>
        <Fonts.Display none>NoNe</Fonts.Display>
        <Fonts.Display uppercase>uppercase</Fonts.Display>
        <Fonts.Display lowercase>LOWERCASE</Fonts.Display>
        <Fonts.Display capitalize>first letters only</Fonts.Display>
      </>
    );

    expect(getByText('NoNe')).toHaveClass('text--none');
    expect(getByText('uppercase')).toHaveClass('text--uppercase');
    expect(getByText('LOWERCASE')).toHaveClass('text--lowercase');
    expect(getByText('first letters only')).toHaveClass('text--capitalize');
  });
});
