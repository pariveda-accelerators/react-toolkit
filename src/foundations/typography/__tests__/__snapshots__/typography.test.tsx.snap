// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography Handles Bold 1`] = `
<body>
  <div>
    <h1
      class="text--display"
      tag="h1"
    >
      <b>
        Bold Display Font
      </b>
    </h1>
  </div>
</body>
`;

exports[`Typography Handles Emphasis 1`] = `
<body>
  <div>
    <h1
      class="text--display"
      tag="h1"
    >
      <em>
        Emphasis Display Font
      </em>
    </h1>
  </div>
</body>
`;

exports[`Typography Handles Italics 1`] = `
<body>
  <div>
    <h1
      class="text--display"
      tag="h1"
    >
      <i>
        Italics Display Font
      </i>
    </h1>
  </div>
</body>
`;

exports[`Typography Handles Strong 1`] = `
<body>
  <div>
    <h1
      class="text--display"
      tag="h1"
    >
      <strong>
        Strong Display Font
      </strong>
    </h1>
  </div>
</body>
`;

exports[`Typography Handles Tag 1`] = `
<body>
  <div>
    <h1
      class="text--display"
      tag="h1"
    >
      H1 Tag
    </h1>
    <h2
      class="text--display"
      tag="h2"
    >
      H2 Tag
    </h2>
    <h3
      class="text--display"
      tag="h3"
    >
      H3 Tag
    </h3>
    <h4
      class="text--display"
      tag="h4"
    >
      H4 Tag
    </h4>
    <h5
      class="text--display"
      tag="h5"
    >
      H5 Tag
    </h5>
    <h6
      class="text--display"
      tag="h6"
    >
      H6 Tag
    </h6>
    <p
      class="text--display"
      tag="p"
    >
      P Tag
    </p>
  </div>
</body>
`;

exports[`Typography Renders Default 1`] = `
<body>
  <div>
    <h1
      class="text--display"
      tag="h1"
    >
      Display Font
    </h1>
    <h2
      class="text--title"
      tag="h2"
    >
      Title Font
    </h2>
    <h3
      class="text--subtitle"
      tag="h3"
    >
      Subtitle Font
    </h3>
    <h4
      class="text--section-title"
      tag="h4"
    >
      SectionTitle Font
    </h4>
    <h5
      class="text--section-subtitle"
      tag="h5"
    >
      Section Subtitle Font
    </h5>
    <p
      class="text--body"
      tag="p"
    >
      Body Font
    </p>
    <p
      class="text--description"
      tag="p"
    >
      Description Font
    </p>
    <h6
      class="text--subsection-title"
      tag="h6"
    >
      Label Font
    </h6>
  </div>
</body>
`;
